cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs

```diff
@@ -94,11 +105,130 @@ import Distribution.Simple.Command
+import qualified Data.ByteString.Char8 as BS
+
+import Network.URI (URI (..), parseURI)
+
+import Distribution.Fields.ConfVar (parseConditionConfVarFromClause)
+
+import Distribution.Client.HttpUtils
+import System.FilePath ((</>), isPathSeparator)
+import System.Directory (createDirectoryIfMissing)
+
+
+
+
+------------------------------------------------------------------
+-- Handle extended project config files with conditionals and imports.
+--
+
+-- | ProjectConfigSkeleton is a tree of conditional blocks and imports wrapping a config. It can be finalized by providing the conditional resolution info
+-- and then resolving and downloading the imports
+type ProjectConfigSkeleton = CondTree ConfVar [ProjectConfigImport] ProjectConfig
+type ProjectConfigImport = String
+
+
+singletonProjectConfigSkeleton :: ProjectConfig -> ProjectConfigSkeleton
+singletonProjectConfigSkeleton x = CondNode x mempty mempty
+
+instantiateProjectConfigSkeleton :: OS -> Arch -> CompilerInfo -> FlagAssignment -> ProjectConfigSkeleton -> ProjectConfig
+instantiateProjectConfigSkeleton os arch impl _flags skel = go $ mapTreeConds (fst . simplifyWithSysParams os arch impl) skel
+    where
+        go :: CondTree
+               FlagName
+               [ProjectConfigImport]
+               ProjectConfig
+             -> ProjectConfig
+        go (CondNode l _imps ts) =
+           let branches = concatMap processBranch ts
+           in l <> mconcat branches
+        processBranch (CondBranch cnd t mf) = case cnd of
+           (Lit True) ->  [go t]
+           (Lit False) -> maybe ([]) ((:[]) . go) mf
+           _ -> error $ "unable to process condition: " ++ show cnd -- TODO it would be nice if there were a pretty printer
+
+projectSkeletonImports :: ProjectConfigSkeleton -> [ProjectConfigImport]
+projectSkeletonImports = view traverseCondTreeC
+
+parseProjectSkeleton :: FilePath -> HttpTransport -> Verbosity -> [ProjectConfigImport] -> FilePath -> BS.ByteString -> IO (ParseResult ProjectConfigSkeleton)
+parseProjectSkeleton cacheDir httpTransport verbosity seenImports source bs = (sanityWalkPCS False =<<) <$> liftPR (go []) (ParseUtils.readFields bs)
+  where
+    go :: [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
+    go acc (x:xs) = case x of
+         (ParseUtils.F l "import" importLoc) ->
+            if importLoc `elem` seenImports
+              then pure . parseFail $ ParseUtils.FromString ("cyclical import of " ++ importLoc) (Just l)
+              else do
+                let fs = fmap (\z -> CondNode z [importLoc] mempty) $ fieldsToConfig (reverse acc)
+                res <- parseProjectSkeleton cacheDir httpTransport verbosity (importLoc : seenImports) importLoc =<< fetchImportConfig importLoc
+                rest <- go [] xs
+                pure . fmap mconcat . sequence $ [fs, res, rest]
+         (ParseUtils.Section l "if" p xs') -> do
+                           subpcs <- go [] xs'
```

Comment:
Why is this block indented by 18 characters, while the previous, analogous block is indented by 3 characters?